pkg/abcupdater/metrics.go

```diff
@@ -0,0 +1,226 @@
+	if lookuper == nil {
+		lookuper = envconfig.OsLookuper()
+	}
+
+	optOutSettings, err := loadOptOutSettings(ctx, lookuper, info.AppID)
+	if err != nil {
+		return fmt.Errorf("failed to load opt out settings: %w", err)
+	}
+
+	if optOutSettings.NoMetrics {
+		return nil
+	}
+
+	generateNewID := true
+	storedID, err := loadInstallID(info)
+	if err == nil && storedID != nil {
+		oneDayAgo := time.Now().Add(-24 * time.Hour)
+		// Defensively check for case ID Created is in future.
+		generateNewID = oneDayAgo.Unix() >= storedID.IDCreatedTimestamp ||
+			time.Now().Unix() < storedID.IDCreatedTimestamp
+	}
+	var installID string
+	if generateNewID {
+		installUUID, err := uuid.NewRandom()
+		if err != nil {
+			return fmt.Errorf("could not generate id for metrics: %w", err)
+		}
+		installID = installUUID.String()
+		err = storeInstallID(info, &InstallIDData{
+			IDCreatedTimestamp: time.Now().Unix(),
+			InstallID:          installID,
+		})
+		if err != nil {
+			logging.FromContext(ctx).DebugContext(ctx, "error storing installID", "error", err.Error())
+		}
+	} else {
+		installID = storedID.InstallID
+	}
+
+	var c metricsConfig
+	if err := envconfig.ProcessWith(ctx, &envconfig.Config{
+		Target:   &c,
+		Lookuper: lookuper,
+	}); err != nil {
+		return fmt.Errorf("failed to process envconfig: %w", err)
+	}
+
+	// Use ParseRequestURI over Parse because Parse validation is more loose and will accept
+	// things such as relative paths without a host.
+	if _, err := url.ParseRequestURI(c.ServerURL); err != nil {
+		return fmt.Errorf("failed to parse server url: %w", err)
+	}
+
+	client := &http.Client{}
+
+	buf := bytes.Buffer{}
+	if err := json.NewEncoder(&buf).Encode(SendMetricRequest{
+		AppID:     info.AppID,
+		Version:   info.Version,
+		Metrics:   info.Metrics,
+		InstallID: installID,
+	}); err != nil {
+		return fmt.Errorf("failed to marshal metric json: %w", err)
+	}
+
+	req, err := http.NewRequestWithContext(ctx, http.MethodPost, fmt.Sprintf(c.ServerURL+"/sendMetrics"), &buf)
+	if err != nil {
+		return fmt.Errorf("failed to create request: %w", err)
+	}
+
+	resp, err := client.Do(req)
+	if err != nil {
+		return fmt.Errorf("failed to make request: %w", err)
+	}
+	defer resp.Body.Close()
+
+	// Future releases may be more strict.
+	if resp.StatusCode >= 300 || resp.StatusCode <= 199 {
+		b, err := io.ReadAll(io.LimitReader(resp.Body, maxErrorResponseBytes))
+		if err != nil {
+			return fmt.Errorf("received %d response, unable to read response body", resp.StatusCode)
+		}
+
+		return fmt.Errorf("received %d response: %s", resp.StatusCode, string(b))
+	}
+
+	// For now, ignore response body for happy responses.
+	// Future versions may parse warnings for debug logging.
+	return nil
+}
+
+// A per-application install id is randomly generated. It gets rotated every
+// 24 hours. This is to avoid a single client polluting metrics.
+func loadInstallID(c *MetricsInfo) (*InstallIDData, error) {
+	path := c.InstallIDFileOverride
```

Comment:
Similar to above, it's strange that this is encapsulated on the `MetricsInfo`; it seems like this would be a testing variable on `MetricsClient`.

Reply:
This has been refactored away